cylc/flow/task_outputs.py

```diff
@@ -70,10 +70,24 @@ def __init__(self, tdef):
         self._by_message = {}
         self._by_trigger = {}
         self._required = set()
+
         # Add outputs from task def.
         for trigger, (message, required) in tdef.outputs.items():
             self._add(message, trigger, required=required)
 
+        # Handle implicit submit requirement
+        if (
+            # "submitted" is not declared as optional/required
+            tdef.outputs[TASK_OUTPUT_SUBMITTED][1] is None
+            # and "submit-failed" is explicitly declared as optional/required
+            and tdef.outputs[TASK_OUTPUT_SUBMIT_FAILED][1] is None
+        ):
```

Comment:
The comments don't match the code here (in the second clause).  The logic seems to work in all cases, from testing, but I'm not quite satisfied that I've thought it through fully yet... [coming back after dinner].

Reply:
I guess the overall logic should be this:
- if neither `:submitted` nor `:submit-failed` are declared (as required or optional) then `:submitted` (required) is implicit and must be added
- otherwise if any of `:submitted` or `:submitted?` or `:submit-failed` or `:submit-failed?` are declared, then the opposite is implied, and there's no need to add implicit `:submitted`

Is that right?

Reply:
I think that's what the logic now does, but the comments need updating.

Reply:
Yes, sorry, the comment is wrong, but the logic is right.

```suggestion
        if (
            # "submitted" is not declared as optional/required
            tdef.outputs[TASK_OUTPUT_SUBMITTED][1] is None
            # and "submit-failed" is not declared as optional/required
            and tdef.outputs[TASK_OUTPUT_SUBMIT_FAILED][1] is None
        ):
```

| submitted | submit-failed | implicit submit dependence |
| --- | --- | --- |
| optional | optional | False |
| optional |  | False |
|  | optional | False |
|  |  | **True** |
| required | required | False |
| required |  | False |
| | required | False |

Note it is not possible to mix "required" and "optional" because the graph parser will raise an error.